### PR TITLE
Structure unification

### PR DESCRIPTION
**Pull request by isbm**
# *\* WORK IN PROGRESS **
### What does this PR do?

Unifies the output from different modules.
### Previous Behavior

`pkg.mod_repo`  returns totally different structures from each other e.g. on RHEL and SLES.
### New Behavior

`pkg.mod_repo` returns identical structures.
### Tests written?

Not yet and probably breaks others. :laughing:

### CODE REVIEW

**Comment by isbm** [comment]
@thatch45 @cachedout @rallytime We have a hard time to bring all the different output to the common (at least lowest) denominator. So I started to do something. At this point, I evaluated various possibilities and at least I would really love to avoid introducing more bugs by correcting/rewriting _existing_ modules that are already working just fine.

What this thing does? Currently I worked on `pkg.mod_repo` and make sure that on RHEL and SLES we get exactly the same output in JSON, so anyone using API do not care the keys to check if they are there, but just know they _are_ there all the time, regardless it is BSD, Debian, SLES, CentOS etc.

This is achieved by an unification set of functions that takes care converting from X to N, where N always outputs the same thing as in constant. Basically, this is the idea to have in Salt something similar to [CIM](https://www.ibm.com/developerworks/ibmi/library/i-cim-ibmi-works/). Except we can still be conventional and expandable, hence pragmatic.

I would like your opinion here.

**Comment by thatch45** [comment]
I know @cachedout will like this approach. I think it is a good idea to make sure these are unified in this way.

**Comment by damon-atkins** [comment]
Would not call the call the class Package as this is targeted at the Repo and the not the packages them selves.  Maybe pkg_repo. ?

**Comment by isbm** [comment]
@damon-atkins From the functional perspective you're right. From object perspective I thought if all of them under the current namespace `structure` then `structure.pkg` is not `modules.pkg`, hence `pkg` on its own is OK. However, you're right: the import should be more explicit. Naming is still subject to change, e.g. I think `scm` is probably a better idea "Salt Common Model". Now I am trying to achieve something that could be similar to like:
```python
from salt.utils.scm import with_scm

@with_scm
def mod_repo(repo, **kwargs):
    ...
```
Basically, a decorator that would figure out by itself what function to what module. Then all what we would need is to import the decorator and decorate all the functions that needs to be unified.

Actually, it is not that bad and not that much functions to unify. Also not everything needs to be unified, but only virtual packages that are same by API interface but different inside the implementation for a different platforms. So I thought situation is much worse, but not that super-scary as much. :smiley:

**Comment by thatch45** [comment]
I like where this is going. But I would like to see what @cachedout thinks as well, he has been thinking about this design for a long time and might have some more ideas.

**Comment by isbm** [comment]
OK, here is one possibility how to deal with the decorators inside the states:
```python
from salt.utils.scm.common import bypass_decorator as bd

def some_cool_state(many, different, *arguments):
    # code
    result = bd(__salt__['unified.output'])(many, different, *arguments)
```

**Comment by isbm** [comment]
Who cached-out @cachedout ? :-) @rallytime in case Mike around, please tell him greetings from a cold Germany. :-)

**Comment by rallytime** [comment]
Haha I sure will! He should be able to look at these changes next week. Thanks @isbm! :)

**Comment by cachedout** [comment]
Hi @isbm. Apologies for the delay here. I was taking a short vacation. :] (And thanks for the kind greetings. It's getting colder in Tokyo now too!)

@thatch45 is right of course; I'm a big fan of this approach. I think that decorators are the right way to go here. I'd like another day or two to think more about this and really give you better feedback because I think we can all agree that we'll want to get this design right before we implement it. I'm going to grab a copy of your fork and play around a bit and get a feel for what you're thinking.

I'd also like to cc: both @terminalmage and @s0undt3ch here as well as @plastikos to get some more discussion going. This is a great start so far!

**Comment by isbm** [comment]
@thatch45 @cachedout I am currently moving forward with this. *Please* take a look at this and give some more suggestions as this kind of needs to be done at some point. Now I just refreshed my infrastructure for all this and did some refactoring.

Thanks!

File: salt/utils/scm/common.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+class Stub(object):
+    '''
+    Class that returns stub message on anything.
+    This shouldn't be happening to the end-user
+    and indicates always a programming error.
+    '''
+
+    def __getattr__(self, item):
+        def stub(*args, **kwargs):
+            return {'error': 'If you got here, your method is missing and you are in nowhere. :)'}
```
Comment (terminalmage) [inline salt/utils/scm/common.py]: This could use a more descriptive message.

**Comment by rallytime** [comment]
Hi @isbm - I just came across this old PR. Is this something you're still working on? Or should it be closed for now?

If this is something you want to keep open, it will need to be rebased against the develop branch. If you want to come back to this another time, we can close it and open it again later. Please let me know what you want to do with this one. :)

**Comment by rallytime** [comment]
I spoke to @isbm on Slack and he said this is good to close as it needs to be re-done anyway.
